### PR TITLE
add Android instrumentation test

### PR DESCRIPTION
**Pull request by yjbanov**
Add an Android instrumentation test. It is similar to `flutter_gallery__transition_perf` but does not require a host machine to run. We can run this test on off-the-shelf test infrastructures, such as Firebase Test Lab or AWS Device Farm.

### CODE REVIEW

File: examples/flutter_gallery/android/app/src/main/java/io/flutter/examples/gallery/FlutterGalleryInstrumentation.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    testFinished = true;
+    result.success(null);
+  }
+
+  public boolean isTestSuccessful() {
+    return testSuccessful;
+  }
+
+  public void waitForTestToFinish() throws Exception {
+    while(!testFinished) {
```
Comment (cbracken) [inline examples/flutter_gallery/android/app/src/main/java/io/flutter/examples/gallery/FlutterGalleryInstrumentation.java]: Uber-nitpick: space between `while` and `(`

**Review by cbracken — APPROVED** [review]

**Comment by cbracken** [comment]
lgtm

File: examples/flutter_gallery/android/app/src/main/java/io/flutter/examples/gallery/FlutterGalleryInstrumentation.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+      Thread.sleep(200);
```
Comment (jason-simmons) [inline examples/flutter_gallery/android/app/src/main/java/io/flutter/examples/gallery/FlutterGalleryInstrumentation.java]: You can avoid polling with a synchronization object such as android.os.ConditionVariable

If you need to poll, then mark the testFinished flag as volatile

Comment (yjbanov) [inline examples/flutter_gallery/android/app/src/main/java/io/flutter/examples/gallery/FlutterGalleryInstrumentation.java]: I'm synchronizing with a Dart isolate, not another Java thread. Will `ConditionVariable` still work?

Marked `testFinished` and `testSuccessful` as `volatile`.

File: examples/flutter_gallery/android/app/src/main/java/io/flutter/examples/gallery/FlutterGalleryInstrumentation.java (hunk 1 above)
Comment (yjbanov) [inline examples/flutter_gallery/android/app/src/main/java/io/flutter/examples/gallery/FlutterGalleryInstrumentation.java]: Done.

File: examples/flutter_gallery/android/app/src/main/java/io/flutter/examples/gallery/FlutterGalleryInstrumentation.java (hunk 2 above)
Comment (yjbanov) [inline examples/flutter_gallery/android/app/src/main/java/io/flutter/examples/gallery/FlutterGalleryInstrumentation.java]: `ConditionVariable` worked! Fixed.

**Review by jason-simmons — APPROVED** [review]

**Pull request merged**